### PR TITLE
Add set() and revert() to Initial plugin

### PR DESCRIPTION
**Pull request by from-nibly**
Implements #208 

This allows you to reset what "initial" is for things like editing forms when you save the form but might want to "edit" and "cancel" after saving. 

Since `set(s)` takes a state value you can use it to set it to any arbitrary state.

The `revert()` method sets the current state to whatever initial was. This could technically be accomplished with  `Initial(state).set(Initial(state).get())`, but would be a handy shortcut.

### CODE REVIEW

File: plugins/Initial/src/Initial.ts [hunk 1]
```diff
@@ -34,6 +36,12 @@ class InitialPluginInstance {
         l.attach(Downgraded)
         return !isEqual(l.value, this.getInitial(l.path))
     }
+    set = (l: StateMethods<StateValueAtRoot>) => {
+        this.initialState = l.value;
```
Comment (avkonst) [inline plugins/Initial/src/Initial.ts:40]: This implementation will be valid only for when `l.path == []`. I believe set and reset should handle setting and resetting nested objects, like set for the `Hookstate.set` does. Have a look how `this.getInitial(l.path)` works and set should do similar to `this.setInitial(l.path)`.

File: plugins/Initial/src/Initial.ts [hunk 2]
```diff
@@ -16,6 +16,8 @@ export interface InitialExtensions<S> {
     get(): S | undefined;
     modified(): boolean;
     unmodified(): boolean;
+    set(s: State<S>): void;
```
Comment (avkonst) [inline plugins/Initial/src/Initial.ts:19]: Once this PR is merged, it would be good to an example to the documentation.

**Review by avkonst — CHANGES_REQUESTED** [review]

File: plugins/Initial/src/Initial.ts (hunk 1 above)
Comment (from-nibly) [reply plugins/Initial/src/Initial.ts:40]: ahh good point. Will do.

File: plugins/Initial/src/Initial.ts (hunk 2 above)
Comment (from-nibly) [reply plugins/Initial/src/Initial.ts:19]: :+1:

**Comment by from-nibly** [comment]
@avkonst made some updates. I found some other issues with the set and revert functions and those are also now fixed.

Unfortunately I'm not sure I have time right at the moment to really test the nested state stuff. It is working for flat state at the moment.

I can probably get around to properly testing it in a few days or so if you want me to make sure that's working before merging it.

File: plugins/Initial/src/Initial.ts [hunk 3]
```diff
@@ -25,15 +26,33 @@ class InitialPluginInstance {
...
-        return !isEqual(l.value, this.getInitial(l.path))
-    }
+        l.attach(Downgraded);
+        return !isEqual(l.value, this.getInitial(l.path));
+    };
+    set = (path: Path, s: StateValueAtPath) => {
+        let result = this.initialState;
+        if (path.length > 0) {
+            for (let i = 0; i < path.length - 1; i++) {
+                result = result && result[path[i]];
```
Comment (avkonst) [inline plugins/Initial/src/Initial.ts:42]: I think it should throw if result is null / undefined here.

Comment (avkonst) [reply plugins/Initial/src/Initial.ts:42]: otherwise, it will modify parent state unexpectedly

File: plugins/Initial/src/Initial.ts [hunk 4]
```diff
@@ -25,15 +26,33 @@ class InitialPluginInstance {
...
+            }
+            result[path[path.length - 1]] = cloneDeep(s);
+        } else {
+            this.initialState = cloneDeep(s);
+        }
+    };
+    revert = (v: StateValueAtRoot) => {
```
Comment (avkonst) [inline plugins/Initial/src/Initial.ts:49]: seems like v is a StateValueAtPath not a StateValueAtRoot

**Review by avkonst — COMMENTED** [review]
nice work, see some of my comments 
and yes, I think it is worth putting automated tests. I do not remember if this plugin had automated tests. If not, you can copy tests infrastructure from some other plugin.

**Pull request closed without merging**